src/org/rascalmpl/shell/ShellEvaluatorFactory.java

```diff
@@ -17,81 +15,85 @@
-        ModuleEnvironment root = heap.addModule(new ModuleEnvironment(ModuleEnvironment.SHELL_MODULE, heap));
-        IValueFactory vf = ValueFactoryFactory.getValueFactory();
-        Evaluator evaluator = new Evaluator(vf, input, stderr, stdout, root, heap, monitor);
-        evaluator.addRascalSearchPathContributor(StandardLibraryContributor.getInstance());
-
-        URIResolverRegistry reg = URIResolverRegistry.getInstance();
+        return getDefaultEvaluator(input, stdout, stderr, monitor, ModuleEnvironment.SHELL_MODULE);
+    }
 
+    public static Evaluator getDefaultEvaluator(Reader input, PrintWriter stdout, PrintWriter stderr, IRascalMonitor monitor, String rootEnvironment) {
+        var reg = URIResolverRegistry.getInstance();
         if (!reg.getRegisteredInputSchemes().contains("project") && !reg.getRegisteredLogicalSchemes().contains("project")) {
-            ISourceLocation rootFolder = PathConfig.inferProjectRoot(URIUtil.rootLocation("cwd"));
-            if (rootFolder != null) {
-                configureProjectEvaluator(evaluator, rootFolder);
-            }
+            return getDefaultEvaluatorForLocation(URIUtil.rootLocation("cwd"), input, stdout, stderr, monitor, rootEnvironment);
         }
+        return getBasicEvaluator(input, stdout, stderr, monitor, rootEnvironment);
+    }
 
-        return evaluator;
+    public static Evaluator getDefaultEvaluatorForPathConfig(PathConfig pcfg, ISourceLocation projectFile, Reader input, PrintWriter stdout, PrintWriter stderr, IRascalMonitor monitor) {
+        return getDefaultEvaluatorForPathConfig(pcfg, projectFile, input, stdout, stderr, monitor, ModuleEnvironment.SHELL_MODULE);
     }
+    
+    public static Evaluator getDefaultEvaluatorForPathConfig(PathConfig pcfg, ISourceLocation projectFile, Reader input, PrintWriter stdout, PrintWriter stderr, IRascalMonitor monitor, String rootEnvironment) {
+        var evaluator = getBasicEvaluator(input, stdout, stderr, monitor, rootEnvironment);
+        var reg = URIResolverRegistry.getInstance();
 
-    public static Evaluator getDefaultEvaluatorForLocation(File fileOrFolderInProject, Reader input, PrintWriter stdout, PrintWriter stderr, IRascalMonitor monitor) {
-        GlobalEnvironment heap = new GlobalEnvironment();
-        ModuleEnvironment root = heap.addModule(new ModuleEnvironment(ModuleEnvironment.SHELL_MODULE, heap));
-        IValueFactory vf = ValueFactoryFactory.getValueFactory();
-        Evaluator evaluator = new Evaluator(vf, input, stderr, stdout, root, heap, monitor);
-        evaluator.addRascalSearchPathContributor(StandardLibraryContributor.getInstance());
+        var projectRoot = PathConfig.inferProjectRoot(projectFile);
+        var projectName = new RascalManifest().getProjectName(projectRoot);
+        reg.registerLogical(new ProjectURIResolver(projectRoot, projectName));
+        reg.registerLogical(new TargetURIResolver(projectRoot, projectName));
```

Comment:
This looks a bit problematic to do for every repl? aspecially as we know that for the REPL in VS Code, this is not correct?

Other cases I worry about: what if the `projectRoot` calculation throws an error or returns null? what if the `projectName` is empty?

Reply:
I think this should maybe be an optional part of the setup? or maybe an extra function of the shell evaluator factor? As this has nothing to do with a path config.